sympy/physics/vector/__init__.py

```diff
@@ -1,40 +1,36 @@
-__all__ = []
-
-# The following pattern is used below for importing sub-modules:
-#
-# 1. "from foo import *".  This imports all the names from foo.__all__ into
-#    this module. But, this does not put those names into the __all__ of
-#    this module. This enables "from sympy.physics.vector import ReferenceFrame" to
-#    work.
-# 2. "import foo; __all__.extend(foo.__all__)". This adds all the names in
-#    foo.__all__ to the __all__ of this module. The names in __all__
-#    determine which names are imported when
-#    "from sympy.physics.vector import *" is done.
-
-from . import frame
-from .frame import *
-__all__.extend(frame.__all__)
-
-from . import dyadic
-from .dyadic import *
-__all__.extend(dyadic.__all__)
-
-from . import vector
-from .vector import *
-__all__.extend(vector.__all__)
-
-from . import point
-from .point import *
-__all__.extend(point.__all__)
-
-from . import functions
-from .functions import *
-__all__.extend(functions.__all__)
-
-from . import printing
-from .printing import *
-__all__.extend(printing.__all__)
-
-from . import fieldfunctions
-from .fieldfunctions import *
-__all__.extend(fieldfunctions.__all__)
+__all__ = [
+    'CoordinateSym', 'ReferenceFrame',
+
+    'Dyadic',
+
+    'Vector',
+
+    'Point',
+
+    'cross', 'dot', 'express', 'time_derivative', 'outer',
+    'kinematic_equations', 'get_motion_params', 'partial_velocity',
+    'dynamicsymbols',
+
+    'vprint', 'vsstrrepr', 'vsprint', 'vpprint', 'vlatex', 'init_vprinting',
+
+    'curl', 'divergence', 'gradient', 'is_conservative', 'is_solenoidal',
+    'scalar_potential', 'scalar_potential_difference',
+
+]
+from .frame import CoordinateSym, ReferenceFrame
+
+from .dyadic import Dyadic
+
+from .vector import Vector
+
+from .point import Point
+
+from .functions import (cross, dot, express, time_derivative, outer,
+        kinematic_equations, get_motion_params, partial_velocity,
+        dynamicsymbols)
+
+from .printing import (vprint, vsstrrepr, vsprint, vpprint, vlatex,
+        init_vprinting)
+
+from .fieldfunctions import (curl, divergence, gradient, is_conservative,
+        is_solenoidal, scalar_potential, scalar_potential_difference)
```

Comment:
Not sure why this way is preferred. The other way lets you manage what is in `__all__` at each module without having to duplicate explicit imports in all subsequent modules.

I'd like it to stay the way we've been writing it here if there is no difference.

Reply:
The advantage of this way is that you can easily see what names are exported by reading `__init__.py`. But we should add some tests that test that everything is consistent across all the files with `__all__`.

Reply:
It isn't necessarily the case that what you get by star-importing from a module should be the same as what you get by star-importing from the containing package so in that sense it isn't duplication.

Probably in most cases we don't care what users get by star-importing from a module and are more concerned with which of those names are being propagate up to package level or to the main sympy namespace. That means that it makes more sense to define the exported names in the `__init__.py` and not bother with `__all__` in the module.

The main points that I see are:
1. If you want a name to propagate out of the module up to say top-level sympy then the exports should be added explicitly at each level so that this is always an explicit decision that contributors and reviewers can see.
2. If a user/reader wants to see what names are exported by a particular package they should be explicitly listed at each level so that you can clearly trace where an import comes from.

Reply:
Are we okay with this style of import in `__init__.py` files? This is just one piece of a larger project that would apply this over the whole of SymPy.

Reply:
I prefer the way it was but will approve for the sake of making sympy consistent.

Reply:
Thanks @moorepants. I'll merge this now